internal/features/stacks/jobs/metadata.go

```diff
@@ -34,13 +36,31 @@ func LoadStackMetadata(ctx context.Context, stackStore *state.StackStore, stackP
 		return err
 	}
 
+	meta, diags := earlydecoder.LoadStack(record.Path(), record.ParsedFiles.AsMap())
+
 	var mErr error
-	meta, diags := earlydecoder.LoadStack(stack.Path(), stack.ParsedFiles.AsMap())
-	if len(diags) > 0 {
-		mErr = diags
+	sErr := stackStore.UpdateMetadata(stackPath, meta, mErr)
+	if sErr != nil {
+		return sErr
 	}
 
-	sErr := stackStore.UpdateMetadata(stackPath, meta, mErr)
+	if len(diags) <= 0 {
+		// no new diagnostics, so return early
+		return mErr
+	}
+
+	// Merge the new diagnostics with the existing ones
+	existingDiags, ok := record.Diagnostics[globalAst.HCLParsingSource]
+	if !ok {
+		existingDiags = make(ast.Diagnostics)
+	} else {
+		existingDiags = existingDiags.Copy()
+	}
+	for fileName, diagnostic := range diags {
+		existingDiags[ast.FilenameFromName(fileName)] = diagnostic
```

Comment:
This will overwrite any existing diagnostics for the file. We could either add a new diagnostics source for the early decoder or implement an `.Extend` method to append them.